### PR TITLE
Add pseudocode for push_front and pop_front

### PR DESCRIPTION
**Pull request by jafingerhut**

### CODE REVIEW

File: p4-16/spec/P4-16-draft-spec.mdk [hunk 1]
```diff
@@ -3363,6 +3363,39 @@ manipulate the elements at the front and back of the stack:
...
+    for (int i = this.size-1; i >= 0; i -= 1) {
+        if (i >= count) {
+            this[i] = this[i-count];
+        } else {
+            this[i].setInvalid();
+        }
+    }
+    this.nextIndex = this.nextIndex + count;
+    if (this.nextIndex > this.size) this.nextIndex = this.size;
+    // Note: this.last, this.next, and this.lastIndex adjust with this.nextIndex
```
Comment (mbudiu-vmw) [inline p4-16/spec/P4-16-draft-spec.mdk]: I don't think we have a lastIndex.

File: p4-16/spec/P4-16-draft-spec.mdk [hunk 2]
```diff
@@ -3363,6 +3363,39 @@ manipulate the elements at the front and back of the stack:
   be a positive integer that is a compile-time known value. The return
   type is ```void```.
 
+The pseudocode below defines the behavior of ```push_front``` and ```pop_front```.
+
+~ Begin P4Pseudo
+void push_front(int count) {
```
Comment (mbudiu-vmw) [inline p4-16/spec/P4-16-draft-spec.mdk]: Can you make count unsigned so we don't have to discuss this issue either?

Comment (jafingerhut) [inline p4-16/spec/P4-16-draft-spec.mdk]: Sure.  I will propagate that change to the other places where push_front and pop_front have and 'int' argument for consistency.  Hopefully that is what you are expecting.

**Comment by mbudiu-vmw** [comment]
In P4 it has to be int, because we don't have unsigned. But in pseudocode it must be positive.
You can at least add a comment that the compiler will reject negative values, and that the values have to be compile-time constants.

File: p4-16/spec/P4-16-draft-spec.mdk (hunk 1 above)
Comment (jafingerhut) [inline p4-16/spec/P4-16-draft-spec.mdk]: There is not a separate data value, but the spec mentions that lastIndex is nextIndex-1.  My comment is there simply to ensure that if some implementer decides to make the .next .last and .lastIndex as separate data items, they should be adjusted as and when nextIndex is updated.

**Comment by jafingerhut** [comment]
Must the language spec restrict the argument of push_front and pop_front to compile time constants?  Can't that be left as an optional target-specific restriction?

**Comment by mbudiu-vmw** [comment]
The problem is that if these are runtime values we have to say what their type is, it must be a bit<N> type for some N.

**Comment by jafingerhut** [comment]
The spec already says that the argument to push_front and pop_front must be positive, and must be compile time constants, a few lines above these proposed additions.  I don't think it should be necessary to repeat it a few lines later.

File: p4-16/spec/P4-16-draft-spec.mdk (hunk 2 above)
Comment (jafingerhut) [inline p4-16/spec/P4-16-draft-spec.mdk]: OK, given your later comment, I am not planning on changing this.  Github still thinks there are changes requested, though.  Not sure how to change that.

**Pull request merged**